crates/sui-node/src/lib.rs

```diff
@@ -641,19 +633,22 @@ impl SuiNode {
     /// This function waits for a signal from the checkpoint executor to indicate that on-chain
     /// epoch has changed. Upon receiving such signal, we reconfigure the entire system.
     pub async fn monitor_reconfiguration(self: Arc<Self>) -> Result<()> {
+        let mut checkpoint_executor = CheckpointExecutor::new(
+            self.state_sync.subscribe_to_synced_checkpoints(),
+            self.checkpoint_store.clone(),
+            self.state.clone(),
+            self.config.checkpoint_executor_config.clone(),
+            &self.registry_service.default_registry(),
+        )?;
+
         loop {
             let Committee {
                 epoch: next_epoch, ..
-            } = self
-                .reconfig_channel
-                .lock()
-                .await
-                .recv()
-                .await
-                .expect("Reconfiguration channel was closed unexpectedly.");
+            } = checkpoint_executor.run_epoch().await;
+
             info!(
                 ?next_epoch,
-                "Received reconfiguration signal. About to reconfigure the system."
+                "Finished executing all checkpoints in epoch. About to reconfigure the system."
```

Comment:
don't we need to do a `send()` on the `end_of_epoch_channel` here?

Reply:
Nope, this PR moves the responsibility of notifying all other components of end of epoch to `SuiNode`. Now, `SuiNode` starts `CheckpointExecutor`, `CheckpointExecutor` returns upon completion with the new committee. Then `SuiNode` does the `send()` to other components (e.g. `TransactionExecutor`), calls reconfig, then restarts all per epoch components (including `CheckpointExecutor`) once the new epoch begins

Reply:
edit: sorry you're absolutely right, I misinterpreted this.